lib/internal/Magento/Framework/Reflection/TypeProcessor.php

```diff
@@ -521,14 +521,149 @@ public function getParamType(ParameterReflection $param)
                 $param->getDeclaringFunction()->name
             ));
         }
-        if ($type == 'array') {
+        if ($type === 'array') {
             // try to determine class, if it's array of objects
             $paramDocBlock = $this->getParamDocBlockTag($param);
             $paramTypes = $paramDocBlock->getTypes();
             $paramType = array_shift($paramTypes);
+
+            $paramType = $this->resolveFullyQualifiedClassName($param->getDeclaringClass(), $paramType);
+
             return strpos($paramType, '[]') !== false ? $paramType : "{$paramType}[]";
         }
-        return $type;
+
+        return $this->resolveFullyQualifiedClassName($param->getDeclaringClass(), $type);
+    }
+
+    /**
+     * Get alias mapping for source class
+     *
+     * @param ClassReflection $sourceClass
+     * @return array
+     */
+    public function getAliasMapping(ClassReflection $sourceClass): array
+    {
+        $sourceFileName = $sourceClass->getDeclaringFile();
+        $aliases = [];
+        foreach ($sourceFileName->getUses() as $use) {
+            if ($use['as'] !== null) {
+                $aliases[$use['as']] = $use['use'];
+            } else {
+                $pos = strrpos($use['use'], '\\');
+
+                $aliasName = substr($use['use'], $pos + 1);
+                $aliases[$aliasName] = $use['use'];
+            }
+        }
+
+        return $aliases;
+    }
+
+    /**
+     * Return true if the passed type is a simple type
+     *
+     * eg.:
+     * Return true with; array, string, ...
+     * Return false with: SomeClassName
+     *
+     * @param string $typeName
+     * @return bool
+     */
+    public function isSimpleType(string $typeName): bool
+    {
+        return strtolower($typeName) === $typeName;
```

Comment:
I can't help but feel like this should be a check against a dictionary of reserved words...

Wouldn't something like `use \Magento\Framework\App\ObjectManager as om;` cause weird problems and error messages that wouldn't make sense?

While it's not the way one _should_ be writing code (subjective), I think we can still prevent the issue if its tackled here?